### PR TITLE
tuxedo-keyboard: fix compilation for kernel 6.10 and 6.11

### PR DESCRIPTION
**Pull request by Keksgesicht**
## Description of changes

It seems like the tuxedo-keyboard has moved to GitLab a little time ago. Thus, I changed the source git the new GitLab URL and updated the version number. This also fixes a bug when compiling with Kernel 6.10.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by K900** [comment]
Please change the commit message format to fit the guidelines in CONTRIBUTING.md

**Comment by K900** [comment]
Also, it may be worth renaming the package to fit its presumably now extended scope?

**Comment by Keksgesicht** [comment]
There are already commit in the new upstream repo to fix compile errors with kernel 6.11. Thus, I will most likely provide a follow up pull request the coming days to update this driver to version 4.6.3.

File: pkgs/top-level/linux-kernels.nix
Comment (MinerSebas) [inline pkgs/top-level/linux-kernels.nix:491]: As you renamed the package here, could you add an alias to block at line 608, in-case someone didn't use the module.
Something like this:
```
tuxedo-keyboard = self.tuxedo-drivers;
```

**Review by MinerSebas — APPROVED** [review]
Still works on my Pulse Gen 1.

**Comment by Keksgesicht** [comment]
Now it also runs with kernel 6.11 on my Aura 15 Gen1.

**Comment by Keksgesicht** [comment]
@blanky0230 What do you think? As you are the maintainer could you review my changes please so it's clear to committers in what state this is?

**Review by blanky0230 — CHANGES_REQUESTED** [review]
Hate to nitpick catching up with reviews this late but: Since the manufacturer's repo is named "tuxedo-drivers" (plural), I think the corresponding linuxPackage and nixosModule in nixpkgs should share the same name in order to minimize the possibility for mix-ups and user's confusion in the future. 

Other than that this PR is looking good to me :+1:

**Comment by papanito** [comment]
@blanky0230 glad you had time to review - some Tuxedo users  like me are waiting for an updated driver 😁

**Comment by blanky0230** [comment]
> @blanky0230 glad you had time to review - some Tuxedo users like me are waiting for an updated driver 😁

Well. All I can do is look at the nix expressions, build the packages, and see whether the modules do as expected since unfortunately I don't have any tuxedo hardware at my disposal at this time. So I can't say 100% the mods do what they're expected to do with the proper hardware.

As stated at: https://github.com/NixOS/nixpkgs/issues/291669#issuecomment-1981588433

I'd love for @CutestNekoAqua to take over the project. @CutestNekoAqua  Maybe you can combine this PR here with your most relevant / urgent updates from https://github.com/NixOS/nixpkgs/pull/293017 and seize the occasion?

**Comment by blanky0230** [comment]
Another trial. https://github.com/NixOS/nixpkgs/pull/343483

Thanks @xaverdh for giving it a spin.

Guess I'll just buy a new tuxedo and join back in to packaging their drivers for nix.

**Comment by Keksgesicht** [comment]
> Hate to nitpick catching up with reviews this late but: Since the manufacturer's repo is named "tuxedo-drivers" (plural)

Not sure, why I kept ingoring the "s". I pushed an updated version.

File: pkgs/os-specific/linux/tuxedo-drivers/default.nix
Comment (blanky0230) [inline pkgs/os-specific/linux/tuxedo-drivers/default.nix:53]: @Keksgesicht @CutestNekoAqua If you like, join yourselves in as maintainers. Certainly better to have a bus-factor > 1 for this package I suppose.
(I'll soon have a suitable device again too).

Comment (blanky0230) [reply pkgs/os-specific/linux/tuxedo-drivers/default.nix]: @Keksgesicht As per #343483 -> https://github.com/NixOS/nixpkgs/blob/2c799a1db3bcc5b1c77cb127d219a1fe8fee864d/pkgs/os-specific/linux/tuxedo-drivers/default.nix#L43

Also getting @xaverdh into the loop. So they know we're trying to tighten up all the various loose ends.

**Comment by Keksgesicht** [comment]
> I'd love for @CutestNekoAqua to take over the project. @CutestNekoAqua Maybe you can combine this PR here with your most relevant / urgent updates from #293017 and seize the occasion?

The most important differences to my PR are the build commands for the kernel module and this is an area I have not enough knowledge. Thus, I would leave my PR in the current state and leave any further changes to @CutestNekoAqua.

File: pkgs/os-specific/linux/tuxedo-drivers/default.nix
Comment (Keksgesicht) [reply pkgs/os-specific/linux/tuxedo-drivers/default.nix]: I am not sure, whether I should add other people. But as they clearly want to develop this further, I could add them in the commit too.

**Comment by blanky0230** [comment]
@CutestNekoAqua Please have a look when you have some time to spare. Also in regards of setting the list of maintainers: Your call. Didn't mean to exclude you from the decision at: https://github.com/NixOS/nixpkgs/pull/336633#discussion_r1779741169

**Review by blanky0230 — APPROVED** [review]

**Comment by CutestNekoAqua** [comment]
> > I'd love for @CutestNekoAqua to take over the project. @CutestNekoAqua Maybe you can combine this PR here with your most relevant / urgent updates from #293017 and seize the occasion?
> 
> The most important differences to my PR are the build commands for the kernel module and this is an area I have not enough knowledge. Thus, I would leave my PR in the current state and leave any further changes to @CutestNekoAqua.

Sure!

File: pkgs/os-specific/linux/tuxedo-drivers/default.nix
Comment (xaverdh) [reply pkgs/os-specific/linux/tuxedo-drivers/default.nix]: feel free to add me!

**Review by xaverdh — COMMENTED** [review]
I would prefer `INSTALL_MOD_PATH` over copying after the fact, but this approach works as well.
Running this on my system now, LGTM

**Review by xaverdh — APPROVED** [review]

**Review by blanky0230 — APPROVED** [review]

**Review by dasj19 — APPROVED** [review]
Changes look good and I have briefly tested in a Gnome environment.
Hotkeys and setting keyboard backlight and fan speed work well.

**Comment by blanky0230** [comment]
@NixOS/nixpkgs-merge-bot merge

**Review by blanky0230 — APPROVED** [review]

**Comment by blanky0230** [comment]
@NixOS/nixpkgs-merge-bot merge

**Comment by WhiteBlackGoose** [comment]
doesn't build for me:
```
error: builder for '/nix/store/l0l897lralncq6mjj8cx1cr0wkjmsm96-tuxedo-keyboard-6.11.2-3.2.14.drv' failed with exit code 2;
       last 10 log lines:
       > /build/source/./src/uniwill_keyboard.h:1259:19: error: initialization of 'void (*)(struct platform_device *)' from incompatible pointer type 'int (*)(struct platform_device *)' [8;;https://gcc.gnu.org/onlinedocs/gcc/W
arning-Options.html#index-Wincompatible-pointer-types-Werror=incompatible-pointer-types8;;]
       >  1259 |         .remove = uniwill_keyboard_remove,
       >       |                   ^~~~~~~~~~~~~~~~~~~~~~~
       > /build/source/./src/uniwill_keyboard.h:1259:19: note: (near initialization for 'platform_driver_uniwill.<anonymous>.remove')
       > cc1: some warnings being treated as errors
       > make[3]: *** [/nix/store/q3jgpnpxqcvh4pcicb9d6ry8mhkwlh9s-linux-6.11.2-dev/lib/modules/6.11.2/source/scripts/Makefile.build:244: /build/source/./src/tuxedo_keyboard.o] Error 1
       > make[2]: *** [/nix/store/q3jgpnpxqcvh4pcicb9d6ry8mhkwlh9s-linux-6.11.2-dev/lib/modules/6.11.2/source/Makefile:1926: /build/source] Error 2
       > make[1]: *** [/nix/store/q3jgpnpxqcvh4pcicb9d6ry8mhkwlh9s-linux-6.11.2-dev/lib/modules/6.11.2/source/Makefile:224: __sub-make] Error 2
       > make[1]: Leaving directory '/nix/store/q3jgpnpxqcvh4pcicb9d6ry8mhkwlh9s-linux-6.11.2-dev/lib/modules/6.11.2/build'
       > make: *** [Makefile:29: all] Error 2
       For full logs, run 'nix log /nix/store/l0l897lralncq6mjj8cx1cr0wkjmsm96-tuxedo-keyboard-6.11.2-3.2.14.drv'.
error: 1 dependencies of derivation '/nix/store/j6wr97kgdks67ypkwb8j1c7p5awpymxh-linux-6.11.2-modules.drv' failed to build
error: 1 dependencies of derivation '/nix/store/b38jgykmzv3chac6yhm5g0cfzhwv3hy0-nixos-system-tuxedo-infinitypro14-nixos-24.11.20241009.5633bcf.drv' failed to build
```
I just copied tuxedo-drivers/default.nix into `tuxedo-new.nix` and added as a boot package:
```nix
  boot.extraModulePackages = [ 
    (config.boot.kernelPackages.callPackage ./tuxedo-new.nix { })
  ];
```

**Comment by Keksgesicht** [comment]
@WhiteBlackGoose. I tested you method myself and it compiles.

Your error message still mentions `tuxedo-keyboard`. You probably did not use the correct nixpkgs commit when you tested this.

In my `flake.nix` I use something like this to test the currently unmerged PR:
```nix
{
  inputs = {
    nixpkgs-master = {
      type = "github";
      owner = "NixOS";
      repo = "nixpkgs";
      ref = "master";
      rev = "c5989b0173fa669297d1f92cee201cf77aa6c3eb";
    };
  };

  outputs = {
    nixpkgs-master,
  }@inputs: {
    nixosConfigurations = {
      "myMachineName" = nixpkgs-master.lib.nixosSystem {
      };
    };
  };
}
```

**Comment by papanito** [comment]
I added the config as @Keksgesicht mentioned in his comment. Build works fine, however somehow I still don't see any backlight on my keyboard.

I also have this
```
  boot = {
    initrd = {
      availableKernelModules = [
        "xhci_pci"
        "thunderbolt"
        "nvme"
        "usb_storage"
        "usbhid"
        "sd_mod"
      ];
      kernelModules = [ ];
    };
    runSize = "50%";
    kernelModules = [ 
      "kvm-intel"
      "tuxedo_keyboard"
    ];
    kernelPackages = pkgs.linuxPackages_latest;
  };  
```

**Comment by Keksgesicht** [comment]
Why haven't you tried using `hardware.tuxedo-drivers.enable` instead? This will also add `boot.extraModulePackages = [ tuxedo-drivers ];`. Making the package which includes the kernel module available to be easily used.

Also, did you reboot after a `nixos-rebuild switch/boot`? Did you try inserting the module into the kernel manually after building?

**Comment by chmanie** [comment]
I would love to help testing this. Is there a way to enable this without setting all of my main config to this branch?

**Comment by papanito** [comment]
> Why haven't you tried using hardware.tuxedo-drivers.enable instead? This will also add boot.extraModulePackages = [ tuxedo-drivers ];. Making the package which includes the kernel module available to be easily used.

I also did try `hardware.tuxedo-drivers.enable`

> Also, did you reboot after a nixos-rebuild switch/boot? Did you try inserting the module into the kernel manually after building

Yes both use the switch parameter and then reboot

**Comment by Keksgesicht** [comment]
@papanito does this show any file:
```bash
realpath /run/*-system/kernel-modules/lib/modules/*/updates/src/tuxedo*
```
or (this is only the case before the latest commit where the build commands changed)
```bash
realpath /run/*-system/kernel-modules/lib/modules/*/tuxedo*
```
Also look whether your kernel loaded the modules:
```bash
lsmod | grep tuxedo
```
Report what is not the case. Did it work with an LTS kernel and the older `tuxedo-keyboard` package?
I am currently unsure if tuxedo has devices which do not need a driver and the backlight is part of the firmware. But there are devices which do not have any backlight.
Also, **without the driver** my device initializes the backlight in blue, but I cannot change the color or brightness. When it goes to sleep/suspend the backlight is gone :(

**Comment by papanito** [comment]
> Report what is not the case. Did it work with an LTS kernel and the older tuxedo-keyboard package?

[It never worked on my InfinityBook](https://discourse.nixos.org/t/installing-tuxedo-keyboard-drivers/40161/4), but I assumed also maybe cause the package seemed very outdated. 

` lsmod` looks fine, `relpath` as well

**Comment by papanito** [comment]
I have 
```
ls /sys/devices/platform/tuxedo_keyboard/leds/
white:kbd_backlight
```

So I can do this

```
echo "51" | sudo tee /sys/devices/platform/tuxedo_keyboard/leds/white:kbd_backlight/brightness
51
```

**Comment by papanito** [comment]
Thanks @Keksgesicht  for your time and effort. I probably have to talk to them yes....

**Comment by chmanie** [comment]
@Keksgesicht thank you, I tested it the way you described. Here are some logs, not sure if this is expected or not (Tuxedo InfinityBook 14 Gen8):

```
| lsmod | rg tuxedo
tuxedo_nb02_nvidia_power_ctrl    12288  0
tuxedo_io              24576  1
tuxedo_keyboard       106496  4 tuxedo_io,uniwill_wmi,clevo_wmi,tuxedo_nb02_nvidia_power_ctrl
tuxedo_compatibility_check    12288  1 tuxedo_keyboard
led_class_multicolor    16384  1 tuxedo_keyboard
battery                28672  2 asus_wmi,tuxedo_keyboard
sparse_keymap          12288  3 intel_hid,asus_wmi,tuxedo_keyboard
led_class              20480  7 snd_hda_codec_generic,videodev,input_leds,iwlmvm,asus_wmi,tuxedo_keyboard,led_class_multicolor
```

And testing tuxedo-rs

```
| systemctl status tailord.service
● tailord.service - Tuxedo Tailor hardware control service
     Loaded: loaded (/etc/systemd/system/tailord.service; enabled; preset: ignored)
     Active: active (running) since Fri 2024-10-18 02:13:59 CEST; 1min 27s ago
 Invocation: ab708c3e7e314e99a79592f56570900b
   Main PID: 1184 (tailord)
         IP: 0B in, 0B out
         IO: 10.9M read, 0B written
      Tasks: 2 (limit: 76804)
     Memory: 6.6M (peak: 7.3M)
        CPU: 294ms
     CGroup: /system.slice/tailord.service
             └─1184 /nix/store/4lnjd6v4gj1wz5idc8g0wwcajq287hqx-tuxedo-rs-0.3.1/bin/tailord

Okt 18 02:13:59 cassis systemd[1]: Starting Tuxedo Tailor hardware control service...
Okt 18 02:13:59 cassis tailord[1184]:  INFO start_runtime: tailord: Starting tailord
Okt 18 02:13:59 cassis tailord[1184]:  INFO start_runtime: tailord::profiles: Loaded profile at `/etc/tailord/active_profile.json`: ProfileInfo { fans: ["default"], leds: [], performance_profile: None }
Okt 18 02:13:59 cassis tailord[1184]:  INFO start_runtime: tailord: Connected to Tuxedo ioctl interface with version 0.3.9
Okt 18 02:13:59 cassis tailord[1184]:  INFO start_runtime: tailord: Tailord started
Okt 18 02:13:59 cassis tailord[1184]:  INFO tailord::suspend: Setting up suspend service
Okt 18 02:13:59 cassis systemd[1]: Started Tuxedo Tailor hardware control service.
Okt 18 02:14:15 cassis tailord[1184]: ERROR tailord::fancontrol: Failed reading the current temperature: `Device not available`
```

**Comment by papanito** [comment]
Lol I feel a bit stupid, but actually it works once I enable the backlight using `Fn`+`Space`

**Comment by Keksgesicht** [comment]
@chmanie looks fine so far. You could test if the `Fn` keys control the `brightness`. If your DM/WM does not comply with that and your InfintyBook has also only a "white" keyboard backlight, you maybe have to do this instead: https://github.com/NixOS/nixpkgs/pull/336633#issuecomment-2419948256

For the failed temperature reading I am not sure whether this driver or a generic one for the Intel or AMD platform provides this sensor. Try running `sensors-detect` of the `lm-sensors` package and look into `/etc/sysconfig/lm_sensors` afterwards. Then add them to `boot.kernelModules` and see whether this ERROR from `tailord` goes away.

**Comment by Keksgesicht** [comment]
So, I rebased and fixed a merge conflict because someone added some system76 things above of my changes in `pkgs/top-level/linux-kernels.nix`. I even updated the package to the newest upstream version.

Is there something else I have to do? How do I get the attention of a committer to merge this?

**Comment by xaverdh** [comment]
The `Failed reading the current temperature` error occurs on startup for me as well, and used to with the old module as well. I think it's race condition with the driver and not something we can fix.

**Comment by xaverdh** [comment]
cc @SuperSandro2000

**Comment by ClockGen** [comment]
@K900 can you please check this PR.

**Comment by WhiteBlackGoose** [comment]
@Keksgesicht I tried what you said (with the new commit):
```nix
    nixpkgs = {
      type = "github";
      owner = "NixOS";
      repo = "nixpkgs";
      ref = "master";
      rev = "fa957fa9d827c2bf471acdf00c89806cec47926b";
    };
```
Sadly I'm getting the same error:
```
error: builder for '/nix/store/ms8133zszndd33if81x0vv8y5d70i8y1-tuxedo-keyboard-6.11.5-3.2.14.drv' failed with exit code 2;
       last 10 log lines:
       > /build/source/./src/uniwill_keyboard.h:1259:19: error: initialization of 'void (*)(struct platform_device *)' from incompatible pointer type 'int (*)(struct platform_device *)' [8;;https://gcc.gnu.org/onlinedocs/gcc/Warning-Options.html#index-Wincompatible-pointer-types-Werror=incompatible-pointer-types8;;]
       >  1259 |         .remove = uniwill_keyboard_remove,
       >       |                   ^~~~~~~~~~~~~~~~~~~~~~~
       > /build/source/./src/uniwill_keyboard.h:1259:19: note: (near initialization for 'platform_driver_uniwill.<anonymous>.remove')
```

**Comment by WhiteBlackGoose** [comment]
setting it directly to your fork worked though:
```nix
    nixpkgs = {
      type = "github";
      owner = "Keksgesicht";
      repo = "nixpkgs";
      ref = "master";
      rev = "8972df5e5ace61612feec3d88b8d768b041dc664";
    };

```

**Review by xaverdh — APPROVED** [review]

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by papanito** [comment]
so now that it's merged to master, how we can use it? Or better when will be in ` unstable`?

**Comment by xaverdh** [comment]
> so now that it's merged to master, how we can use it? Or better when will be in ` unstable`?

The unstable channel just advanced a few hours ago, so it should be there already.

**Comment by xaverdh** [comment]
by the way thanks for merging this @SuperSandro2000 :heart:

**Pull request merged**
